File: src/typing/flow_cache.ml
(*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *)

open Utils_js
open Reason
open Type
open TypeUtil
module ImplicitTypeArgument = Instantiation_utils.ImplicitTypeArgument

(* Cache that remembers pairs of types that are passed to __flow. *)
module FlowConstraint = struct
  let rec toplevel_use_op = function
    | Frame (_frame, use_op) -> toplevel_use_op use_op
    | Op (Speculation use_op) -> toplevel_use_op use_op
    | use_op -> use_op

  (* attempt to read LB/UB pair from cache, add if absent *)
  let get cx (l, u) =
    match (l, u) with
    (* Don't cache constraints involving type variables, since the
       corresponding typing rules are already sufficiently robust. *)
    | (OpenT _, _)
    | (_, UseT (_, OpenT _))
    | (_, ReposUseT _) ->
      false
    | _ ->
      (* Use ops are purely for better error messages: they should have no
         effect on type checking. However, recursively nested use ops can pose
         non-termination problems. To ensure proper caching, we hash use ops
         to just their toplevel structure. *)
      let u = mod_use_op_of_use_t toplevel_use_op u in
      let cache = Context.constraint_cache cx in
      let cache' = FlowSet.add l u !cache in
      let found = cache' == !cache in
      if not found then
        cache := cache'
      else if Context.is_verbose cx then
        prerr_endlinef
          "%sFlowConstraint cache hit on (%s, %s)"
          (Context.pid_prefix cx)
          (string_of_ctor l)
          (string_of_use_ctor u);
      found
end

(* Cache that limits instantiation of polymorphic definitions. Intuitively,
   for each operation on a polymorphic definition, we remember the type
   arguments we use to specialize the type parameters. An operation is
   identified by its reason, and possibly the reasons of its arguments. We
   don't use the entire operation for caching since it may contain the very
   type variables we are trying to limit the creation of with the cache (e.g.,
   those representing the result): the cache would be useless if we considered
   those type variables as part of the identity of the operation. *)
module PolyInstantiation = struct
  let find cx reason_tapp typeparam op_reason =
    let cache = Context.instantiation_cache cx in
    let in_synthesis_mode = Context.in_synthesis_mode cx in
    match
      Reason.ImplicitInstantiationReasonMap.find_opt
        (reason_tapp, typeparam.reason, op_reason, in_synthesis_mode)
        !cache
    with
    | Some t -> t
    | None ->
      let t = ImplicitTypeArgument.mk_targ cx typeparam (Nel.hd op_reason) reason_tapp in
      cache :=
        Reason.ImplicitInstantiationReasonMap.add
          (reason_tapp, typeparam.reason, op_reason, in_synthesis_mode)
          t
          !cache;
      t
end

module Eval = struct
  let id cx t defer_use =
    let (eval_id_cache, id_cache) = Context.eval_id_cache cx in
    match t with
    | EvalT (_, d, i) when d = defer_use ->
      (match Type.EvalIdCacheMap.find_opt i !eval_id_cache with
      | Some t -> t
      | None ->
        let i = Type.Eval.generate_id () in
        eval_id_cache := Type.EvalIdCacheMap.add i t !eval_id_cache;
        EvalT (t, defer_use, i))
    | _ ->
      let cache_key = (t, defer_use) in
      let id =
        match Type.IdCacheMap.find_opt cache_key !id_cache with
        | Some i -> i
        | None ->
          let i = Type.Eval.generate_id () in
          id_cache := Type.IdCacheMap.add cache_key i !id_cache;
          i
      in
      EvalT (t, defer_use, id)

  let find_repos cx t defer_use id =
    let repos_cache = Context.eval_repos_cache cx in
    let cache_key = (t, defer_use, id) in
    Type.EvalReposCacheMap.find_opt cache_key !repos_cache

  let add_repos cx t defer_use id tvar =
    let repos_cache = Context.eval_repos_cache cx in
    let cache_key = (t, defer_use, id) in
    repos_cache := Type.EvalReposCacheMap.add cache_key tvar !repos_cache
end

module Fix = struct
  let find cx is_this i =
    let cache = Context.fix_cache cx in
    let cache_key = (is_this, i) in
    Type.FixCacheMap.find_opt cache_key !cache

  let add cx is_this i tvar =
    let cache = Context.fix_cache cx in
    let cache_key = (is_this, i) in
    cache := Type.FixCacheMap.add cache_key tvar !cache
end

(* debug util: please don't dead-code-eliminate *)
(* Summarize flow constraints in cache as ctor/reason pairs, and return counts
   for each group. *)
let summarize_flow_constraint cx =
  let group_counts =
    FlowSet.fold
      (fun (l, u) map ->
        let key =
          spf
            "[%s] %s => [%s] %s"
            (string_of_ctor l)
            (string_of_reason (reason_of_t l))
            (string_of_use_ctor u)
            (string_of_reason (reason_of_use_t u))
        in
        match SMap.find_opt key map with
        | None -> SMap.add key 0 map
        | Some i -> SMap.add key (i + 1) map)
      !(Context.constraint_cache cx)
      SMap.empty
  in
  SMap.elements group_counts |> List.sort (fun (_, i1) (_, i2) -> Stdlib.compare i1 i2)

File: rust_port/crates/flow_typing_flow_common/src/flow_cache.rs
/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

use std::collections::BTreeMap;

use flow_typing_context::Context;
use flow_typing_type::type_::Type;
use flow_typing_type::type_::TypeDestructorT;
use flow_typing_type::type_::TypeInner;
use flow_typing_type::type_::UseOp;
use flow_typing_type::type_::UseT;
use flow_typing_type::type_::UseTInner;
use flow_typing_type::type_::VirtualRootUseOp;
use flow_typing_type::type_::VirtualUseOp;
use flow_typing_type::type_::eval as type_eval;
use flow_typing_type::type_::string_of_ctor;
use flow_typing_type::type_::string_of_use_ctor;
use flow_typing_type::type_util::mod_use_op_of_use_t;
use flow_typing_type::type_util::reason_of_t;
use flow_typing_type::type_util::reason_of_use_t;

// Cache that remembers pairs of types that are passed to __flow.
pub mod flow_constraint {
    use std::ops::Deref;

    use dupe::Dupe;

    use super::*;

    fn toplevel_use_op(use_op: &UseOp) -> UseOp {
        match &use_op {
            VirtualUseOp::Frame(_, inner_use_op) => toplevel_use_op(inner_use_op),
            VirtualUseOp::Op(root) => {
                if let VirtualRootUseOp::Speculation(inner) = &**root {
                    toplevel_use_op(inner)
                } else {
                    use_op.dupe()
                }
            }
        }
    }

    /// Returns `true` iff `toplevel_use_op` would return the same UseOp
    /// unchanged (i.e. it is already an `Op(non-Speculation root)`). The
    /// common case for non-speculation callers — checking this lets the
    /// outer cache lookup skip the `mod_use_op_of_use_t` allocation that
    /// would otherwise produce a structurally-identical `UseT`.
    fn is_toplevel_use_op(use_op: &UseOp) -> bool {
        match use_op {
            VirtualUseOp::Frame(_, _) => false,
            VirtualUseOp::Op(root) => !matches!(&**root, VirtualRootUseOp::Speculation(_)),
        }
    }

    // attempt to read LB/UB pair from cache, add if absent
    pub fn get<'cx>(cx: &Context<'cx>, l: &Type, u: &UseT<Context<'cx>>) -> bool {
        match (l.deref(), u.deref()) {
            // Don't cache constraints involving type variables, since the
            // corresponding typing rules are already sufficiently robust.
            (TypeInner::OpenT(_), _) => false,
            (_, UseTInner::UseT(_, inner_t)) if matches!(inner_t.deref(), TypeInner::OpenT(_)) => {
                false
            }
            (_, UseTInner::ReposUseT(..)) => false,
            // Don't cache concretization constraints, since the concretization operation should already
            // have defense against bad cyclic types.
            (_, UseTInner::ConcretizeT(..)) => false,
            (
                _,
                UseTInner::BindT(..)
                | UseTInner::CallT(..)
                | UseTInner::MethodT(..)
                | UseTInner::PrivateMethodT(..)
                | UseTInner::ConstructorT(..)
                | UseTInner::LookupT(..)
                | UseTInner::GetPropT(..)
                | UseTInner::ObjRestT { .. }
                | UseTInner::ObjTestT { .. }
                | UseTInner::ArrRestT(..)
                | UseTInner::ValueToTypeReferenceT(..)
                | UseTInner::SpecializeT(..)
                | UseTInner::EvalTypeDestructorT(..)
                | UseTInner::ConcretizeTypeAppsT { .. }
                | UseTInner::ExtendsUseT(..)
                | UseTInner::ResolveUnionT(..)
                | UseTInner::ConditionalT(..)
                | UseTInner::DeepReadOnlyT { .. }
                | UseTInner::ElemT(..)
                | UseTInner::ReposLowerT { .. }
                | UseTInner::ObjKitT { .. }
                | UseTInner::HasOwnPropT(..),
            ) => false,
            _ => {
                // Use ops are purely for better error messages: they should have no
                // effect on type checking. However, recursively nested use ops can pose
                // non-termination problems. To ensure proper caching, we hash use ops
                // to just their toplevel structure.
                //
                // Fast path: if every use_op the `UseT` carries is already
                // top-level, `mod_use_op_of_use_t` would allocate a fresh
                // `Rc<UseTInner>` whose contents are bit-for-bit identical
                // to the input. Skip the allocation in that case and reuse
                // `u.dupe()` (cheap Rc clone) for both the cache key and
                // the verbose-log path.
                let already_toplevel = match u.deref() {
                    UseTInner::UseT(use_op, _) => is_toplevel_use_op(use_op),
                    _ => false,
                };
                let u_owned = if already_toplevel {
                    u.dupe()
                } else {
                    mod_use_op_of_use_t(toplevel_use_op, u)
                };
                let mut cache = cx.constraint_cache_mut();
                let found = !cache.add(l.dupe(), u_owned.dupe());
                if !found {
                    // New entry
                } else if cx.is_verbose() {
                    eprintln!(
                        "{}FlowConstraint cache hit on ({}, {})",
                        cx.pid_prefix(),
                        string_of_ctor(l),
                        string_of_use_ctor(&u_owned)
                    );
                }
                found
            }
        }
    }
}

pub mod eval {
    use dupe::Dupe;

    use super::*;

    pub fn id(cx: &Context, t: Type, defer_use: TypeDestructorT) -> Type {
        let mut eval_id_cache = cx.eval_id_cache_mut();
        let mut id_cache = cx.id_cache_mut();

        if let TypeInner::EvalT {
            defer_use_t: d,
            id: i,
            ..
        } = &*t
        {
            if d == &defer_use {
                if let Some(cached_t) = eval_id_cache.get(i) {
                    return cached_t.dupe();
                }
                let i = type_eval::Id::generate_id();
                eval_id_cache.insert(i.dupe(), t.dupe());
                return Type::new(TypeInner::EvalT {
                    type_: t,
                    defer_use_t: defer_use,
                    id: i,
                });
            }
        }

        let cache_key = (t.dupe(), defer_use.dupe());
        let id = match id_cache.get(&cache_key).cloned() {
            Some(i) => i,
            None => {
                let i = type_eval::Id::generate_id();
                id_cache.insert(cache_key, i.dupe());
                i
            }
        };
        Type::new(TypeInner::EvalT {
            type_: t,
            defer_use_t: defer_use,
            id,
        })
    }

    pub fn find_repos(
        cx: &Context,
        t: &Type,
        defer_use: &TypeDestructorT,
        id: &type_eval::Id,
    ) -> Option<Type> {
        let repos_cache = cx.eval_repos_cache();
        let cache_key = (t.dupe(), defer_use.dupe(), id.dupe());
        repos_cache.get(&cache_key).map(|t| t.dupe())
    }

    pub fn add_repos(
        cx: &Context,
        t: Type,
        defer_use: TypeDestructorT,
        id: type_eval::Id,
        tvar: Type,
    ) {
        let mut repos_cache = cx.eval_repos_cache_mut();
        let cache_key = (t, defer_use, id);
        repos_cache.insert(cache_key, tvar);
    }
}

pub mod fix {
    use dupe::Dupe;

    use super::*;

    pub fn find(cx: &Context, is_this: bool, i: &Type) -> Option<Type> {
        let cache = cx.fix_cache();
        let cache_key = (is_this, i.dupe());
        cache.get(&cache_key).map(|t| t.dupe())
    }

    pub fn add(cx: &Context, is_this: bool, i: Type, tvar: Type) {
        let mut cache = cx.fix_cache_mut();
        let cache_key = (is_this, i);
        cache.insert(cache_key, tvar);
    }
}

// debug util: please don't dead-code-eliminate
// Summarize flow constraints in cache as ctor/reason pairs, and return counts
//    for each group.
pub fn summarize_flow_constraint(cx: &Context) -> Vec<(String, i32)> {
    let cache = cx.constraint_cache();
    let mut group_counts: BTreeMap<String, i32> = BTreeMap::new();

    cache.fold(
        |(), (l, u)| {
            let key = format!(
                "[{}] {:?} => [{}] {:?}",
                string_of_ctor(l),
                reason_of_t(l),
                string_of_use_ctor(u),
                reason_of_use_t(u)
            );
            *group_counts.entry(key).or_insert(0) += 1;
        },
        (),
    );

    let mut result: Vec<(String, i32)> = group_counts.into_iter().collect();
    result.sort_by(|(_, i1), (_, i2)| i1.cmp(i2));
    result
}
